puppet/lib/puppet/util/puppetdb/char_encoding.rb

```diff
@@ -104,13 +86,14 @@ def self.error_char_context(str, bad_char_indexes)
   # @param error_context_str information about where this string came from for use in error messages
   # @return String
   def self.warn_if_invalid_chars(str, error_context_str)
-    bad_char_indexes = all_indexes_of_char(str, DEFAULT_INVALID_CHAR)
-    if bad_char_indexes.empty?
+
+    if str.index(DEFAULT_INVALID_CHAR).nil?
       str
     else
       Puppet.warning "#{error_context_str} ignoring invalid UTF-8 byte sequences in data to be sent to PuppetDB, see debug logging for more info"
+
       if Puppet.settings[:log_level] == "debug"
-        Puppet.debug error_context_str + "\n" + error_char_context(str, bad_char_indexes).join("\n")
+        Puppet.debug error_context_str + "\n" + error_char_context(str, first_invalid_char_range(str))
```

Comment:
The new code only outputs debugging info for the first chunk of invalid characters. This overhead will only occur when debug mode is enabled.